macro/src/lib.rs

```diff
@@ -194,6 +209,24 @@ fn gen_params_invocation(param: &ast::Param, expanded_params: &mut Vec<Expr>) {
             let param = &param.name;
             expanded_params.push(parse2(quote!(#param.into())).unwrap());
         }
+        ast::TypeName::Function(in_types, out_type) => {
+            let cb_wrap_ident = Ident::new(&format!("{}_cb_wrap", param.name), Span::call_site());
+            let mut cb_param_list = vec![];
+            let mut cb_arg_type_list = vec![];
+            for (index, in_ty) in in_types.into_iter().enumerate() {
+                cb_param_list.push(Ident::new(&format!("arg{}", index), Span::call_site()));
+                cb_arg_type_list.push(in_ty.to_syn());
+            }
+            let cb_ret_type = out_type.to_syn();
+
+            let tokens = quote! {
+                move | #(#cb_param_list,)* | unsafe {
```

Comment:
one thing I'm a bit unsure of: I was under the impression we performed some amount of type conversion between the FFI type and the Rust type in the macro

specifically:

 - slices are passed over FFI as a `data, len` pair of parameters
 - we call `from_raw_parts_mut()` etc to make that work
 - we also convert `Result` to `DiplomatResult` where needed (this is not yet relevant here)

I think we need to be handling the slices here, similarly, yes?

Though at this stage perhaps it's better for Diplomat to just uniformly use the DiplomatFooView types (cc @robertbastian)? This avoids the issue of splitting things into multiple args, quite cleanly.

Reply:
We'd still need to call `.into()`, but it would save us the trouble of splatting the arguments

Reply:
Filed https://github.com/rust-diplomat/diplomat/issues/613. Not a problem for you now, I might refactor a bunch of the macro code if I fix #613, which I think i will

Reply:
Ah that would be helpful, thanks! I'll refactor to fit your changes when you fix that.

Reply:
After #615, "types needing conversion" is much simpler to think about.


This can be handled in a followup which we should file when we merge this. There are two routes here:

1. We can forbid types that need conversion in callback args. All types that need conversion have a pure-FFI version that looks uglier (potentially) but is perfectly usable
2. We teach the callback macro codegen code to check for types-needing-conversion and call .into() when passing things to the callback


(2) is preferable, but we can go with (1) initially. For this PR it's fine if we just don't perform conversions for now (it will cause compilation failures if you write callbacks that take types-that-need-conversion (Rust slices, later, `Option<T>` for non-opaque T).